src/drivers/intel/cavs/ssp.c

```diff
@@ -37,6 +37,55 @@ static int hweight_32(uint32_t mask)
 	return count;
 }
 
+/**
+ * \brief Finds valid M/(N * SCR) values for given frequencies.
+ * \param[in] freq SSP clock frequency.
+ * \param[in] bclk Bit clock frequency.
+ * \param[out] out_scr_div SCR divisor.
+ * \param[out] out_m M value of M/N divider.
+ * \param[out] out_n N value of M/N divider.
+ * \return true if found suitable values, false otherwise.
+ */
+static bool find_mn(uint32_t freq, uint32_t bclk,
+		    uint32_t *out_scr_div, uint32_t *out_m, uint32_t *out_n)
+{
+	uint32_t m, n, mn_div;
+	uint32_t scr_div;
+
+	/* M/(N * scr_div) has to be less than 1/2 */
+	if ((bclk * 2) >= freq)
+		return false;
+
+	scr_div = freq / bclk;
+
+	/* odd SCR gives lower duty cycle */
+	if (scr_div > 1 && scr_div % 2 != 0)
+		--scr_div;
+
+	/* clamp to valid SCR range */
+	scr_div = MIN(scr_div, (SSCR0_SCR_MASK >> 8) + 1);
+
+	/* find highest even divisor */
```

Comment:
@singalsu Do you know better way to find it? It can take ~2000 iterations in pessimistic scenario. Maybe you know some clever match hack to find out it more efficiently.

Reply:
thats an ultra rare theoretical case, in real scenario this while will even not start so its ok